cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java

```diff
@@ -0,0 +1,134 @@
+                                                            Set<Integer> excludedBrokers,
+                                                            Replica replica,
+                                                            ActionType action) {
+    if (!excludedBrokers.isEmpty()
+        && (action == ActionType.LEADERSHIP_MOVEMENT || (replica.originalBroker().isAlive() && replica.isLeader()))) {
+      originalBrokers.removeIf(broker -> excludedBrokers.contains(broker.id()));
+    }
+  }
+
+  /**
+   * Filter the given candidate brokers in the given clusterModel to retrieve the eligible ones for execution of a
+   * {@link com.linkedin.kafka.cruisecontrol.analyzer.ActionType#REPLICA_MOVEMENT} or
+   * {@link com.linkedin.kafka.cruisecontrol.analyzer.ActionType#LEADERSHIP_MOVEMENT} action for the given replica.
+   *
+   * Invariant-1: If there are new brokers, an eligible candidate that triggers an action must be a new broker.
+   * Invariant-2: Brokers excluded for leadership if exclusion applies to the given replica for the given action.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param replica  Replica to check for action eligibility.
+   * @param candidates Candidate brokers among which the eligible ones will be selected.
+   * @param action Action that affects the given replica.
+   * @param optimizationOptions Options to take into account while applying the given action.
+   * @return List of eligible brokers with a fixed order.
+   */
+  static List<Broker> eligibleBrokers(ClusterModel clusterModel,
+                                      Replica replica,
+                                      Collection<Broker> candidates,
+                                      ActionType action,
+                                      OptimizationOptions optimizationOptions) {
+
+    List<Broker> eligibleBrokers = new ArrayList<>(candidates);
+    filterOutBrokersExcludedForLeadership(eligibleBrokers, optimizationOptions.excludedBrokersForLeadership(), replica, action);
+
+    if (clusterModel.newBrokers().isEmpty()) {
+      return eligibleBrokers;
+    }
+
+    // When there are new brokers, we should only allow the replicas/leadership to be moved to the new brokers.
+    return eligibleBrokers.stream().filter(b -> b.isNew() || b == replica.originalBroker()).collect(Collectors.toList());
+  }
+
+  /**
+   * Check whether the proposed action is legit. An action is legit if it is:
+   * (1) a replica movement and the destination does not have a replica of the same partition, or
+   * (2) a leadership movement, the replica is a leader and the destination broker has a follower of the same partition.
+   *
+   * @param replica Replica that is affected from the given action type.
+   * @param destinationBroker Destination broker.
+   * @param actionType Action type.
+   * @return True if the move is legit, false otherwise.
+   */
+  static boolean legitMove(Replica replica, Broker destinationBroker, ActionType actionType) {
```

Comment:
can we add a sanity check in this method to double check that no leader is moved into excluded broker. Because the logic to filter out the excluded broker from candidate broker is scatter in code. Through it looks like we add filtering in all the place for now, it is possible in the future that in some new operation-generating code we miss this filtering.

Reply:
Adding the proposed sanity check to `legitMove` may have some issues:
1. The check would be incomplete due to goals that do not use `legitMove` -- e.g. `PreferredLeaderElectionGoal`.
2. `legitMove` is a frequently accessed method -- i.e. it is used for each balancing action proposal. Hence, adding an extra logic here leads to a significant overhead.
3. Ideally, our unit tests should catch such violations. To mitigate this issue, I added test cases that we missed before.